Improve support for event-driven IO

hanno-arm opened the pull request:
__Summary:__ This PR enhances our support for event-driven IO both in the library and on the testing side, the need for which was triggered through work on the issue underlying PR #772 (IOTSSL-1204).

__Test infrastructure:__ 
* A `pack=TIME` option for our UDP proxy allowing to merge multiple records into a single UDP datagram.
* An 'UDP proxy wrapper' which silently puts a server binary like `ssl_server2` behind a UDP proxy. This is useful e.g. to run our test-suites with UDP packing added everywhere without the need to change server application or test script.
* Add `event=1` option to `ssl_server2` and `ssl_client2` to enable event-driven IO. With that, the applications idle when receiving a `WANT_READ` or `WANT_WRITE`, until they either receive an event from the timer or the underlying transport.

__Library:__
* Add a new public polling function for `mbedtls_net_context_t` allowing to wait for events on mbed TLS provided network contexts.
* Add a new public function `mbedtls_ssl_check_pending` to check whether the internal buffers still contain data to be processed. This is an inevitable addition because a user employing event driven IO needs that information to decide when he/she can idle on the underlying transport. Note that this is different from `mbedtls_ssl_get_bytes_avail` which only concerns application data. A simple usecase demonstrating the need for `mbedtls_ssl_check_pending` is the case when multiple application data records are contained in a single UDP datagram. Having fully read the data from the first datagram, `mbedtls_ssl_get_bytes_avail` will return `0`, but idling on the underlying transport might deadlock.

More information is in the commit messages.

__Internal reference:__ IOTSSL-1204

__Note__ While without the changes of this PR it's not possible to do any testing around event-driven IO, in particular IOTSSL-1204, it's not mandatory for them to get merged to close IOTSSL-1204. The particular bugfix for IOTSSL-1204 will be done in a separate PR.

__Please review__

gilles-peskine-arm commented on library/ssl_tls.c:
```diff
@@ -3526,81 +3540,16 @@ static int ssl_parse_record_header( mbedtls_ssl_context *ssl )
-     * version, length) are sane.
+     * DTLS-related tests.
+     * Must be done before checking the bounds for the length
+     * field because the latter depends on the current epoch.
```
I don't see where that dependency is when reading the code. Could you clarify which length depends on the epoch and how?

hanno-arm replied:
@gilles-peskine-arm See IOTSSL-1417. The minimal record length depends on the negotiated cipher suite and hence changes with the epoch; during the first handshake, it's `1`, afterwards it's larger because of the presence of a MAC e.g. The reason I swapped the code-paths is that records with an old epoch are silently discarded without skipping the datagram, while records that are too short lead to dismissal of the entire datagram. This leads to a problem in the following laboratory test-case: The `ChangeCipherSpec` message from the first handshake got duplicated and merged with the subsequent `Finished` message. After the first `ChangeCipherSpec`, the minimal record length increases to `> 1`, which renders the duplicated `ChangeCipherSpec` too short and therefore leads to dismissal of the entire datagram, including the `Finished` message. In contrast, swapping the length and epoch checks leads to the duplicated `ChangeCipherSpec` being silently dropped, and the `Finished` message is parsed as expected.

It's a long time ago that I worked on this PR. Perhaps we should consult the DTLS RFC to see if it mandates a specific order of checks, and whether such strange laboratory scenarios as the one just described are potentially excluded.

I came across that issue while playing with the UDP proxy and UDP packer - duplication and packing in conjunction lead to the above problem quite quickly.

gilles-peskine-arm replied:
Thanks for the explanation, please put it in a comment.

hanno-arm replied:
@mpg I didn't find anything in the standard saying that some failures should lead to dropping an entire datagram, while others should only lead to discarding the record. The only relevant section seems to be [this one](https://tools.ietf.org/html/rfc6347#section-4.1.2.7). Do you know more about that?

hanno-arm replied:
@gilles-peskine-arm Done.

gilles-peskine-arm commented on library/ssl_tls.c:
```diff
@@ -3640,9 +3589,74 @@ static int ssl_parse_record_header( mbedtls_ssl_context *ssl )
             return( MBEDTLS_ERR_SSL_UNEXPECTED_RECORD );
         }
```
Not necessarily related to this patch, but: how do we know that there are enough bytes in the input buffer for `mbedtls_ssl_dtls_replay_check`? It reads 8 bytes from `ssl->in_ctr`.

hanno-arm replied:
@gilles-peskine-arm `ssl_parse_record_header` is called [from here](https://github.com/ARMmbed/mbedtls/blob/development/library/ssl_tls.c#L3873), so at least `13` record header bytes are available in case of DTLS, and the 8 bytes starting at `ssl->in_ctr` [belong to that](https://github.com/ARMmbed/mbedtls/blob/development/library/ssl_tls.c#L5635).

gilles-peskine-arm replied:
Good, thanks for checking!

gilles-peskine-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -801,26 +897,39 @@ int main( int argc, char *argv[] )
         {
-            mbedtls_printf("forced ciphersuite not allowed with this protocol version\n");
+            mbedtls_printf( "forced ciphersuite not allowed with "
+                            "this protocol version\n" );
```
Please don't break up string literals. People grep for them.

hanno-arm replied:
@gilles-peskine-arm Honestly, I have done that in multiple places throughout other PR's, too, in order to adhere to our style requirements. Not sure what we want to give preference here.

gilles-peskine-arm replied:
@hanno-arm Our coding standards state that 80 characters is the “preferred” maximum length, and if there ever was a good reason to break this general rule, it's to fit string literals on a single line. Even the extremely picky rules for the Linux kernels forbid breaking debug messages.

mpg replied:
I think historically we used to break string literals in order to fit in the 80 characters limit, but I agree with Gilles that we shouldn't. We might want to edit the style document to clarify that.

hanno-arm replied:
@mpg @gilles-peskine-arm Given that there'll be many occasions where this needs to be changed, I'd prefer to keep this PR as is and to postpone documenting and implementing this style requirement to a separate PR. Would you be ok with that?

gilles-peskine-arm replied:
@hanno-arm I don't expect this PR to fix any preexisting split debugging messages, but please don't introduce any new one.

gilles-peskine-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -397,6 +403,77 @@ static int my_verify( void *data, mbedtls_x509_crt *crt, int depth, uint32_t *fl
 }
 #endif /* MBEDTLS_X509_CRT_PARSE_C */
 
+/*
```
That's quite a bit of code duplicated between `ssl_server2.c` and `ssl_client2.c`. Should we introduce an `ssl_util2.c`?

hanno-arm replied:
We already have other duplications between `ssl_server2.c` and `ssl_client2.c`, such as `my_debug`, `my_send` and `my_recv`. If we want to outsource these similarities, I'd prefer we open a separate ticket for it.

gilles-peskine-arm replied:
Ok: IOTSSL-1807

hanno-arm commented on include/mbedtls/net_sockets.h:
```diff
@@ -132,6 +137,27 @@ int mbedtls_net_accept( mbedtls_net_context *bind_ctx,
+ * \param timeout  Maximal amount of time to wait before returning,
+ *                 in milliseconds. If \c timeout is zero, the
+ *                 function returns immediately. A negative value
+ *                 allows the function to block indefinitely.
```
Ok.

hanno-arm replied:
@gilles-peskine-arm Done.

hanno-arm commented on include/mbedtls/net_sockets.h:
```diff
@@ -132,6 +137,27 @@ int mbedtls_net_accept( mbedtls_net_context *bind_ctx,
+ * \param ctx      Socket to check
+ * \param rw       Bitflag composed of MBEDTLS_NET_POLL_READ and
+ *                 MBEDTLS_NET_POLL_WRITE specifying the checks to
+ *                 perform.
```
Thanks, I think I messed up here, expecting `select` to return only if both read- and write-descriptor were ready (if set), which is wrong. As you say, the way it is now the return value isn't very useful. I'll change it.

gilles-peskine-arm replied:
@hanno-arm `select` returns as soon as at least one of the channels is ready. That's the whole point of the function — if you want to wait for all the channels to be ready, you can wait for each of them successively (assuming that ready channels don't spontaneously become unready).

hanno-arm replied:
@gilles-peskine-arm Yes, thanks for pointing that out!

hanno-arm replied:
@gilles-peskine-arm Yes, thanks for pointing that out! Done.

hanno-arm commented on include/mbedtls/net_sockets.h:
```diff
@@ -132,6 +137,27 @@ int mbedtls_net_accept( mbedtls_net_context *bind_ctx,
+ *                 allows the function to block indefinitely.
+ *
+ * \return         0 if all checks are successful, 1 if not, or
+ *                 MBEDTLS_ERR_NET_POLL_FAILED on failure.
```
Yes, see above, it doesn't make sense now, I misread the documentation of `select`.

hanno-arm replied:
Done.

hanno-arm commented on library/net_sockets.c:
```diff
@@ -434,6 +434,56 @@ int mbedtls_net_set_nonblock( mbedtls_net_context *ctx )
+    fd_set read_fds;
+    fd_set write_fds;
+
+    int fd = ((mbedtls_net_context *) ctx)->fd;
```
Done.

hanno-arm commented:
@gilles-peskine-arm I think I have addressed your comments. Could you have another look please? I'm not yet a bash ninja, so if you could in particular check whether there's some room for improvement in `udp-proxy-wrapper.sh`, I'd be happy.

hanno-arm commented:
@gilles-peskine-arm I have rebased and cleaned up the history of this PR.

mpg commented on programs/ssl/ssl_client2.c:
```diff
@@ -433,6 +437,78 @@ static int ssl_sig_hashes_for_test[] = {
+     * if data is still pending to be processed. */
+    if( mbedtls_ssl_check_pending( ssl ) != 0 )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "WARNING: Data still pending, "
```
Again, I feel like this belongs to the "should not happen" category and should result in an error.

mpg commented on programs/ssl/ssl_client2.c:
```diff
@@ -433,6 +437,78 @@ static int ssl_sig_hashes_for_test[] = {
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "WARNING: Data still pending, "
+                                    "but idling requested!" ) );
+    }
+    MBEDTLS_SSL_DEBUG_MSG( 3, ( "idle, waiting for event... " ) );
```
Note: It somehow feels weird to me to use `MBEDTLS_SSL_DEBUG_MSG` from the application code, as opposed to the library SSL modules. OTOH, I can't name a single practical concrete issue with it, so perhaps it's just me.

mpg commented on library/ssl_tls.c:
```diff
@@ -6392,6 +6392,67 @@ size_t mbedtls_ssl_get_bytes_avail( const mbedtls_ssl_context *ssl )
+     */
+
+    /* TODO This needs correction in the same way as
+     *      read_record_layer, see IOTSSL-1414 */
```
As discussed in another PR, please clearly mention this is an internal reference.

hanno-arm commented on programs/test/udp_proxy.c:
```diff
@@ -560,14 +662,30 @@ int main( int argc, char *argv[] )
+        ctx_buffer_check( &outbuf[0] );
+        ctx_buffer_check( &outbuf[1] );
+
+        if( ( ret = select( nb_fds, &read_fds, NULL, NULL, &tm ) ) < 0 )
```
Here `select` is actually used to poll on multiple file descriptors simultaneously, something that `mbedtls_net_poll` isn't able to do.

mpg replied:
But I'm not convinced the simultaneity is critical, it seems to me we could replace each of the subsequent `FD_ISSET()` with a direct call to `mbedtls_net_poll()`. Anyway, that's tangential and not important for acceptance of the PR.

hanno-arm replied:
@mpg Sorry, I mistakenly remembered that this was a blocking call to `select`, but it's not - and even if it had been, since we're in a loop here anyway we can use the non-blocking version of `net_poll` sequentially for client and server.

gilles-peskine-arm replied:
I just realized that this is a busy-wait loop. Please don't do that, even on a host-only program. The use of `select` is justified here, but pass an infinite timeout.

hanno-arm replied:
@gilles-peskine-arm @mpg If we want to pass an infinite timeout, then we have to stick to `select`, though, and cannot use `mbedtls_net_poll`.

hanno-arm replied:
@gilles-peskine-arm As far as I see, we're busy-waiting in `ssl_server2` and `ssl_client2`, too.

mpg replied:
I think the timeout should not be infinite, but computed as the minimum of the "wait to see if there are other packets to pack with the current" times of the two buffer contexts.

(Btw I agree the use of `select()` is justified then.)

hanno-arm commented on tests/ssl-opt.sh:
```diff
@@ -597,16 +597,22 @@ fi
 # sanity checks, avoid an avalanche of errors
-if [ ! -x "$P_SRV" ]; then
-    echo "Command '$P_SRV' is not an executable file"
+P_SRV_BIN=$(echo "$P_SRV" | sed -r -n "s/^([^ ]*).*$/\1/p")
```
Yes, @gilles-peskine-arm remarked the same :+1: and it's already adapted in #946.

hanno-arm commented on programs/test/udp_proxy.c:
```diff
@@ -152,6 +157,7 @@ static void get_options( int argc, char *argv[] )
     opt.server_port    = DFL_SERVER_PORT;
     opt.listen_addr    = DFL_LISTEN_ADDR;
     opt.listen_port    = DFL_LISTEN_PORT;
+    opt.merge          = DFL_PACK;
```
@mpg Thanks, I'll use `opt.pack` instead.

hanno-arm commented on programs/test/udp_proxy.c:
```diff
@@ -108,6 +109,8 @@ int main( void )
     "\n"                                                                    \
     "    seed=%%d             default: (use current time)\n"                \
+    "    pack=%%d             default: 0     (don't merge)\n"               \
+    "                         options: t > 0 (merge for t milliseconds)\n"  \
```
@mpg Thanks, I have rewritten the code to use the portable code from `timing.h` instead.

gilles-peskine-arm commented:
Subsumed by #946 which has been approved by reviewers. Closing, we won't merge this one unless we decide that #946 isn't ready to release after all. Note that if we reopen, Manuel and I need to finish reviewing, some of the rework of #946 may be needed here.

gilles-peskine-arm commented:
#946 is now merged, so this can stay closed.
